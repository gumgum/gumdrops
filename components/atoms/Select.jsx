import React from 'react';
import PropTypes from 'prop-types';
import trimString from '../utils/trimString';

const Select = ({ context, size, customValue, customName, options, className, ...otherProps }) => {
    const baseClass = 'gds-form-group__select-input',
        contextClass = context ? `${baseClass}--${context}` : '',
        sizeClass = size ? `${baseClass}--${size}` : '';

    const classNames = trimString(`${baseClass} ${contextClass} ${sizeClass} ${className}`);

    return (
        <select className={classNames} {...otherProps}>
            {options.map(o => (
                <option key={o[customValue]} value={o[customValue]}>
                    {o[customName]}
                </option>
            ))}
        </select>
    );
};

Select.displayName = 'Select';

Select.defaultProps = {
    options: [],
    customValue: 'value',
    customName: 'name'
};

Select.propTypes = {
    /** primary, no-border, dark */
    context: PropTypes.string,
    /** xs, sm, lg */
    size: PropTypes.oneOf(['xs', 'sm', 'lg']),
    customValue: PropTypes.string,
    customName: PropTypes.string,
    options: PropTypes.array,
    className: PropTypes.string
};

export default Select;
