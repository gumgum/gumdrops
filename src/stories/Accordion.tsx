import React from 'react';
import cx from 'classnames';

interface AccordionCommonProps {
    color?: 'dark' | 'white';
    size?: 'sm' | '';
}

type OmitColor<T> = Omit<T, 'color'>;

export interface AccordionItemProps
    extends AccordionCommonProps,
        OmitColor<React.LiHTMLAttributes<HTMLLIElement>> {
    isOpen?: boolean;
    isLocked?: boolean;
    label: React.ReactChild;
}

export interface AccordionProps
    extends AccordionCommonProps,
        OmitColor<React.HTMLAttributes<Element>> {
    initialAllOpen?: boolean;
}

export const Accordion: React.FC<AccordionProps> = ({
    children,
    color,
    size = '',
    className,
    initialAllOpen,
    ...otherProps
}) => {
    const rootClass = cx('gds-accordion', className, {
        [`gds-accordion--${color}`]: color
    });

    const newChildren = React.Children.map(children, child => {
        if (!React.isValidElement<AccordionItemProps>(child)) {
            return child;
        }

        return React.cloneElement<AccordionItemProps>(child, {
            color,
            size,
            isOpen: child.props.isOpen ? child.props.isOpen : initialAllOpen
        });
    });

    return (
        <div className={rootClass} {...otherProps}>
            <ul className="gds-accordion-list">{newChildren}</ul>
        </div>
    );
};
