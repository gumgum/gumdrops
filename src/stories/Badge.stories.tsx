import React from 'react';
// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import { Story, Meta } from '@storybook/react/types-6-0';

import { Badge, BadgeProps } from './Badge';
import { Colors } from '../types';

export default {
    title: 'Badge',
    component: Badge
} as Meta;

const Template: Story<BadgeProps> = args => <Badge {...args} />;

export const Primary = Template.bind({});
Primary.args = {
    color: Colors.primary,
    text: 'Primary badge'
};

export const Warning = Template.bind({});
Warning.args = {
    color: Colors.warning,
    text: 'Warning Badge'
};

export const Empty = Template.bind({});
Empty.args = {
    color: Colors.info,
    empty: true
};
