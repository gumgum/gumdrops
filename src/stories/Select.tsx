import React, { ChangeEventHandler } from 'react';
import cx from 'classnames';
import { Colors, SelectOptions, Sizes } from 'types';

export interface SelectPropsInterface extends React.SelectHTMLAttributes<HTMLSelectElement> {
    onChange: (value: unknown) => void;
    options: SelectOptions;
    color?: Colors;
    customValue?: string;
    customName?: string;
}

export type SelectProps = Omit<SelectPropsInterface, 'size'> & {
    size?: Sizes;
};

const baseClass = 'gds-form-group__select-input';

export const Select: React.FC<SelectProps> = ({
    onChange,
    color,
    size,
    customValue = 'value',
    customName = 'name',
    options,
    className,
    disabled,
    ...otherProps
}) => {
    const rootClass = cx(baseClass, className, {
        [`${baseClass}--${color}`]: color,
        [`${baseClass}--${size}`]: size,
        '-disabled': disabled,
        '-pointer-events--none': disabled
    });

    const getChange: ChangeEventHandler = event => {
        const value = (event.currentTarget as HTMLInputElement).value;
        onChange(value);
    };

    return (
        <select onChange={getChange} className={rootClass} disabled={disabled} {...otherProps}>
            {options.map(option => {
                const isObject = typeof option === 'object' && option !== null;
                const key = isObject ? option[customName] : option;
                const value = isObject ? option[customValue] : option;
                return (
                    <option key={key} value={value}>
                        {key}
                    </option>
                );
            })}
        </select>
    );
};
