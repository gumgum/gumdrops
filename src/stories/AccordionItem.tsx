import React, { MouseEvent, KeyboardEvent, useState } from 'react';
import cx from 'classnames';
import { AccordionItemProps } from './Accordion';
import { keyCodes } from 'types';

type ToggleEvent = MouseEvent<HTMLLIElement> | KeyboardEvent<HTMLLIElement>;

export const AccordionItem: React.FC<AccordionItemProps> = ({
    isOpen: initialIsOpen,
    isLocked,
    label,
    size,
    color,
    className,
    children
}) => {
    const [isOpen, setIsOpen] = useState(initialIsOpen);

    const toggleOpen = (event: ToggleEvent): void => {
        event.stopPropagation(); // don't want nested accordions to propagate events
        event.preventDefault(); // prevents page scroll from space key
        const { type } = event;
        const { key } = event as KeyboardEvent;

        if (isLocked) return;
        const allowToggle =
            (type === 'keypress' && [keyCodes.space, keyCodes.enter].includes(key as keyCodes)) ||
            type === 'click';

        if (!allowToggle) return;

        setIsOpen(isOpen => !isOpen);
    };

    const rootClass = cx('gds-accordion__item', className, {
        [`gds-accordion__item--${color}`]: color,
        'gds-accordion__item--active': isOpen
    });

    const titleClass = cx('gds-accordion__item-title', {
        [`gds-accordion__item-title--${size}`]: size
    });

    const iconClass = cx('gds-accordion__item-icon', {
        [`gds-accordion__item-icon--${size}`]: size
    });

    const childClass = cx('gds-accordion__child-items', {
        [`gds-accordion__child-items--${size}`]: size
    });

    const newChildren = React.Children.map(children, child => {
        if (!React.isValidElement<AccordionItemProps>(child)) {
            return child;
        }

        return React.cloneElement<AccordionItemProps>(child, {
            color,
            size
        });
    });

    const regionId = `AccordionItem-region-${label}`;
    const labelId = `AccordionItem-label-${label}`;

    return (
        <li
            aria-pressed={isOpen}
            aria-controls={regionId}
            aria-expanded={isOpen}
            className={rootClass}
            onClick={toggleOpen}
            onKeyPress={toggleOpen}
            role="button"
            tabIndex={isLocked ? -1 : 0}>
            <h4 id={labelId} className={titleClass}>
                {label}
            </h4>
            {!isLocked && <span className={iconClass} />}
            <ul
                aria-labelledby={labelId}
                aria-hidden={!isOpen}
                className={childClass}
                id={regionId}
                role="region">
                {newChildren}
            </ul>
        </li>
    );
};
