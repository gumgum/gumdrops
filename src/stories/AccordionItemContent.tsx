import React from 'react';
import cx from 'classnames';
import { AccordionItemProps } from './Accordion';

export const AccordionItemContent: React.FC<Omit<AccordionItemProps, 'label'>> = ({
    className,
    color,
    size,
    children,
    ...otherProps
}) => {
    const stopClick: React.MouseEventHandler = e => {
        e.stopPropagation();
    };

    const rootClass = cx('gds-accordion__child-item', className, {
        [`gds-accordion__child-item--${color}`]: color
    });

    const titleClass = cx('gds-accordion__child-item-title', {
        [`gds-accordion__child-item-title--${size}`]: size
    });

    return (
        <li className={rootClass} onClick={stopClick} {...otherProps}>
            <h4 className={titleClass}>{children}</h4>
        </li>
    );
};
