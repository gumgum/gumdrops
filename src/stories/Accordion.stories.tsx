import React from 'react';
// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import { Story, Meta } from '@storybook/react/types-6-0';

import { Accordion, AccordionProps } from './Accordion';
import { AccordionItem } from './AccordionItem';
import { AccordionItemContent } from './AccordionItemContent';

export default {
    title: 'Accordion',
    component: Accordion
} as Meta;

const Template: Story<AccordionProps> = args => (
    <Accordion {...args}>
        <AccordionItem label="I'm locked and open" isLocked isOpen>
            <AccordionItemContent>Locked Content</AccordionItemContent>
        </AccordionItem>
        <AccordionItem label="Item 1">
            <AccordionItemContent>Content 1</AccordionItemContent>
        </AccordionItem>
        <AccordionItem label="Item 2">
            <AccordionItemContent>Content 2</AccordionItemContent>
        </AccordionItem>
        <AccordionItem label="Item 3">
            <AccordionItemContent>Content 3</AccordionItemContent>
            <Accordion>
                <AccordionItem label="Nested Item">
                    <AccordionItemContent>Nested Content</AccordionItemContent>
                </AccordionItem>
            </Accordion>
        </AccordionItem>
    </Accordion>
);

export const Primary = Template.bind({});
Primary.args = {};
