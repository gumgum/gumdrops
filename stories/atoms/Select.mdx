import { Canvas, Source } from '@storybook/blocks';
import { ArgsTable } from '@storybook/addon-docs';

import * as SelectStories from './Select.stories';
import Select from '../../components/atoms/Select';

# Select

The `<Select>` component is used to allow the user to choose a value from a set of options. It is recommended that you wrap it with the `<FormGroup>` component.

## Options

The default format for each option is `{ name: 'my name', value: 'my value' }`. If your options are in a different format, you may pass in `customValue` and/or `customName` props to override the default values of "name" and "value". For example, if your data looks like `{ id: 'my id', val: 'my value' }`, you can pass in `customName="id"` and `customValue="val"`.

## Controlled vs Uncontrolled

This component can be controlled or uncontrolled. This component accepts any prop that you may want to add. The only drawback is that there is not validation for unlisted props. **It is highly recommended that you use controlled components in React.**

_Recommended Props_: `name`, `value`, `defaultValue`, `onChange`

### Controlled example

```jsx
const fruitOptions = [
    { name: 'bananas', value: 1 },
    { name: 'apples', value: 2 }
];

const [selectedOption, setSelectedOption] = useState(1);
return (
    <FormGroup>
        <FormGroupLabel text="Fruits" />
        <Select
            options={fruitOptions}
            name="fruit"
            value={selectedOption}
            onChange={e => {
                setSelectedOption(e.target.value);
            }}
        />
    </FormGroup>
);
```

### Uncontrolled example

```jsx
_handleSubmit = ({ target }) => console.log(target.fruit.value);

render() {
    return (
        <form onSubmit={ this._handleSubmit }>
            <FormGroup>
                <FormGroupLabel text="Fruits" />
                <Select
                   options={ myOptions }
                   name="fruit"
                   defaultValue="3"
                />
            </FormGroup>
        </form>
    );
}
```

## Component Example

<Canvas of={SelectStories.Default} />

## Component Example Source

<Source of={SelectStories.Default} />

## Select Props

<ArgsTable of={Select} />
