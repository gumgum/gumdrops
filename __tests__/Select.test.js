import React from 'react';
import { render, screen } from '@testing-library/react';
import user from '@testing-library/user-event';
import { Select } from '../src/stories/Select';

describe('Select', () => {
    it('changes the count select', () => {
        const mockOnChange = jest.fn();
        const options = [
            { name: 'first', value: 1, foo: 'foo1' },
            { name: 'second', value: 'Second', foo: 'foo2' },
            3
        ];
        const { container } = render(
            <Select options={options} onChange={mockOnChange} value={'Second'} />
        );
        screen.getByDisplayValue('second');
        user.selectOptions(screen.getByRole('combobox'), '3');
        expect(mockOnChange).toHaveBeenCalledWith('3');
        user.selectOptions(screen.getByRole('combobox'), 'first');
        expect(mockOnChange).toHaveBeenCalledWith('1');
        expect(container.firstChild).toMatchSnapshot();
    });

    it('does nothing when disabled', () => {
        const mockOnChange = jest.fn();
        const options = [
            { name: 'first', value: 1, foo: 'foo1' },
            { name: 'second', value: 'Second', foo: 'foo2' },
            3
        ];
        render(<Select options={options} onChange={mockOnChange} value={'Second'} disabled />);
        user.selectOptions(screen.getByRole('combobox'), '3');
        expect(mockOnChange).not.toHaveBeenCalled();
    });
});
