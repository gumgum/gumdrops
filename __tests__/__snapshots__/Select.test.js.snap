// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select changes the count select 1`] = `
<select
  class="gds-form-group__select-input"
>
  <option
    value="1"
  >
    first
  </option>
  <option
    value="Second"
  >
    second
  </option>
  <option
    value="3"
  >
    3
  </option>
</select>
`;
